File: ml/src/main/scala/ai/h2o/sparkling/ml/features/H2OWord2VecTokenizer.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ai.h2o.sparkling.ml.features

import ai.h2o.sparkling.ml.params.H2OWord2VecTokenizerParams
import org.apache.spark.annotation.DeveloperApi
import org.apache.spark.ml.Transformer
import org.apache.spark.ml.feature.{RegexTokenizer, StopWordsRemover}
import org.apache.spark.ml.param._
import org.apache.spark.ml.util.{DefaultParamsReadable, DefaultParamsWritable, Identifiable}
import org.apache.spark.sql.expressions.UserDefinedFunction
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types.{StringType, StructField, StructType}
import org.apache.spark.sql.{DataFrame, Dataset}

/**
  * Tokenize Data so they are ready for H2OWord2Vec
  */
class H2OWord2VecTokenizer(override val uid: String)
  extends Transformer
  with H2OWord2VecTokenizerParams
  with DefaultParamsWritable {

  def this() = this(Identifiable.randomUID(getClass.getSimpleName))

  @DeveloperApi
  override def transformSchema(schema: StructType): StructType = {
    require(getInputCol != null, "Input column has to be specified!")
    StructType(schema.fields ++ Array(StructField(getOutputCol(), StringType)))
  }

  private def addValue: UserDefinedFunction = udf((array: Seq[String]) => array ++ Array(""))

  override def transform(dataset: Dataset[_]): DataFrame = {
    require(getInputCol != null, "Input column has to be specified!")
    val tokenizer = new RegexTokenizer()
      .setInputCol(getInputCol())
      .setMinTokenLength(getMinTokenLength())
      .setOutputCol(s"tokenized_$uid")
      .setPattern(getPattern())

    val stopWordsRemover = new StopWordsRemover()
      .setInputCol(tokenizer.getOutputCol)
      .setStopWords(getStopWords())
      .setOutputCol(s"stop_words_removed_$uid")

    val tokenized = tokenizer
      .transform(dataset)
      .withColumn(tokenizer.getOutputCol, addValue(col(tokenizer.getOutputCol)))

    stopWordsRemover
      .transform(tokenized)
      .withColumn(getOutputCol(), explode(col(stopWordsRemover.getOutputCol)))
      .drop(tokenizer.getOutputCol, stopWordsRemover.getOutputCol)
  }

  override def copy(extra: ParamMap): Transformer = defaultCopy(extra)
}

object H2OWord2VecTokenizer extends DefaultParamsReadable[H2OWord2VecTokenizer]

File: py/src/ai/h2o/sparkling/ml/features/H2OWord2VecTokenizer.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from ai.h2o.sparkling import Initializer
from ai.h2o.sparkling.ml.H2OStageBase import H2OStageBase
from ai.h2o.sparkling.ml.Utils import Utils
from ai.h2o.sparkling.ml.params.H2OTypeConverters import H2OTypeConverters
from pyspark import keyword_only
from pyspark.ml.param.shared import *
from pyspark.ml.wrapper import JavaTransformer


class H2OWord2VecTokenizer(H2OStageBase, JavaTransformer):
    stopWords = Param(
        Params._dummy(),
        "stopWords",
        "Remove specified words from the array vector in the dataset",
        H2OTypeConverters.toNullableListString())

    inputCol = Param(
        Params._dummy(),
        "inputCol",
        "Input Column",
        H2OTypeConverters.toNullableString())

    outputCol = Param(
        Params._dummy(),
        "outputCol",
        "Output Column",
        H2OTypeConverters.toNullableString())

    @keyword_only
    def __init__(self,
                 stopWords=False,
                 inputCol=None,
                 outputCol=None):
        Initializer.load_sparkling_jar()
        super(H2OWord2VecTokenizer, self).__init__()
        self._java_obj = self._new_java_obj("ai.h2o.sparkling.ml.features.H2OWord2VecTokenizer", self.uid)
        self._setDefaultValuesFromJava()
        kwargs = Utils.getInputKwargs(self)
        self._set(**kwargs)

    def setStopWords(self, value):
        return self._set(stopWords=value)

    def setInputCol(self, value):
        return self._set(inputCol=value)

    def setOutputCol(self, value):
        return self._set(outputCol=value)

    def getStopWordds(self):
        return self.getOrDefault(self.stopWords)

    def getInputCol(self):
        return self.getOrDefault(self.inputCol)

    def getOutputCol(self):
        return self.getOrDefault(self.outputCol)
